spec/mem/tree_borrows/tree.md

```diff
@@ -0,0 +1,140 @@
+# Tree Structure in Tree Borrows
+
+The core data structure of the Tree Borrows is a *Tree*. We use a tree to track reborrows in each allocation. The tree consists of several nodes, each corresponding to a borrow tag (i.e., a reborrow operation). Each node has a parent and a list of children. Additionally, on memory locations (bytes) within the allocation, we track each tag's state defined in [state_machine.md](state_machine.md).
+
+```rust
+struct Node {
+    children: List<Node>,
+    /// State for each location
+    location_states: List<LocationState>,
+}
+```
+
+During each memory access, we update states according to the state machine defined in [state_machine.md](state_machine.md). When a node is accessed, each node in the tree can be divided into two disjoint sets: *Child* and *Foreign*, based on its relative position to the accessed node. The Child set includes the node itself and all its descendants, while the Foreign set contains all other nodes.
+
+```rust
+enum NodeRelation {
+    Child, 
+    Foreign,
+}
+```
+
+The state transition depends on both the node relation and the type of access operation.
+
+```rust
+enum AccessKind {
+    Read, 
+    Write,
+}
+```
+
+Then we implement how to actually update and check states in a Tree during each memory access.
+
+```rust
+impl Node {
+    /// Perform state transition on all locations of `self`.
+    /// The `node_relation` indicates whether the accessed node is a child or foreign
+    /// *from the perspective of the `self`*.
+    fn transition(
+        &mut self, 
+        node_relation: NodeRelation,
+        access_kind: AccessKind,
+        offset_in_alloc: Size,
+        size: Size
+    ) -> Result {
+        let offset_start = offset_in_alloc.bytes();
+        for offset in offset_start..offset_start + size.bytes() {
+            let curr_state = self.location_states[offset];
+            let next_state = curr_state.transition(access_kind, node_relation)?;
+            self.location_states.set(offset, next_state);
+        }
+
+        ret(())
+    }
+
+    /// Recusively do state transition on the `self` node and all its descendants.
+    /// The `path` is the path from the `self` to the accessed node.
+    /// The `is_ancestor` indicates whether the `self` is an ancestor of the accessed node.
+    ///
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn access(
+        &mut self,
+        path: Path, // root -> child
+        access_kind: AccessKind,
+        offset_in_alloc: Size,
+        size: Size,
+        is_ancestor: bool,
```

Comment:
I am confused by this, it seems redundant with `path`? Specifically, if there is a path from `self` to the accessed node, then `self` must be an ancestor of the accessed node?

Would it work to have an `Option<Path>`, where `None` means "accessed node is not a descendant of `self`"?